### ISSUE TITLE
Object Detection (Coral) stopping without any error

### ISSUE THREAD

**Issue by arrikhan** [codeproject/CodeProject.AI-ObjectDetectionCoral#10]
**Area of Concern**
- [ ] Server
CodeProjectAI 2.8.0 - Docker build
- [ ] Behaviour 
Object Detection (Coral) is stopping without any logging.
- [ ] Installer
docker image: codeproject/ai-server:2.8.0

**Describe the bug**
When opening web UI, every few days I see the Coral module has stopped. I press restart and it happily starts up and starts processing again. 

**Expected behavior**
I expect the module to keep running and/or attempt automatic restart. I cannot find any configuration to tell it to autorestart.

**Your System (please complete the following information):**
 - CodeProject.AI Server version: 2.8.0
 - OS: Ubuntu 24.04.1 LTS
 - System RAM 8Gb
 - GPU N/A (running a mac mini 2012)
 - GPU N/A 
 - Coral TPU (USB)

**Comment by ChangoMunk** [comment]
> **How this was put together:** I worked through the source with Claude, which drafted the analysis and the patch below. The observations, logs and crash reports are all from my own machine, and the code claims are straightforwardly checkable against `main` — but the patch has not been compiled or tested by either of us, so please treat it as an untested proposal rather than a ready change. Where I'm inferring rather than reporting something I've observed, I've said so.

I'm hitting this on **2.9.5 on Windows** with a Coral USB Accelerator and Blue Iris, so it isn't Docker- or Linux-specific. I went digging through the source to work out why there's no auto-restart, and I think the cause is in the server rather than in the Coral module. Findings and a proposed patch below.

## Symptom

The Coral object detection module runs normally for hours or days, then stops. Detection latency is healthy right up to the moment it dies — no degradation, no warning:

```
15:36:36:Response rec'd from Object Detection (Coral) command 'detect' (...3ebcde) ['No objects found']  took 67ms
15:36:36:Response rec'd from Object Detection (Coral) command 'detect' (...1b0673) ['No objects found']  took 64ms
15:37:24:Module ObjectDetectionCoral has shutdown
15:37:24:objectdetection_coral_adapter.py: has exited
```

It then stays stopped indefinitely — in my case for about 18 hours before I noticed. The dashboard shows "Stopped", and nothing attempts to bring it back. Restarting it by hand from the dashboard starts the module again.

Nothing in the log distinguishes this from a shutdown I asked for, which is why it's easy to miss.

## Root cause

Two gaps, both in the server. Analysed against `main` @ `e3468c8`, where `src/server/version.json` reads **2.9.6** — so this is current code, not an old release.

### 1. There's no supervision after startup

`ModuleRunner` (`src/server/Modules/ModuleRunner.cs`) is a `BackgroundService`. `ExecuteAsync` registers routes, starts each module once, installs initial modules, and then:

```csharp
try
{
    await Task.Delay(Timeout.Infinite, stoppingToken).ConfigureAwait(false);
}
catch (TaskCanceledException)
{
}
```

It sleeps until shutdown. Nothing re-checks whether the modules it started are still alive. Grepping the server for `BackgroundService` / `IHostedService` / `PeriodicTimer` returns only `MeshMonitor`, `ServerVersionProcessRunner`, `ModuleRunner` itself, and one unrelated `BackgroundServiceExceptionBehavior.Ignore` in host setup — nothing that watches module processes.

So the only thing that can notice a module dying is the `Process.Exited` event, which leads to:

### 2. The exit handler can't tell a crash from a shutdown

`ModuleProcessServices.ModuleExecutionComplete` is wired up as the `Exited` handler for every module process, in `StartProcess`:

```csharp
process.Exited += ModuleExecutionComplete;
```

Its own doc comment says it covers both cases — "This is called once the module's execution has completed (shutdown or crash)" — but the body doesn't branch on which happened:

```csharp
TryGetProcessStatus(moduleId, out ProcessStatus? processStatus);
if (processStatus is not null)
    processStatus.Status = ProcessStatusType.Stopped;

_logger.LogInformation($"** Module {moduleId} has shutdown");

// Remove this from the list of running processes
if (_runningProcesses.TryGetValue(moduleId, out _))
    _runningProcesses.TryRemove(moduleId, out _);
```

Every exit — a deliberate stop, a deliberate restart, a segfault in a native library, an OOM kill — produces the same `Stopped` status, the same `LogInformation` line, and no recovery attempt. The exit code isn't read or logged.

That's why the original report here says "stopping without any logging". The crash *is* logged, but it's logged as though it were a clean shutdown, at information level.

### The design looks like it anticipated this

`ProcessStatusType.Crashed` is defined in `src/SDK/NET/Modules/ProcessStatus.cs`, and `dashboard.js` already renders it:

```javascript
case "Crashed":
    statusDesc = "Crashed";
    className = "alert-danger"
    break;
```

But `grep -rn "ProcessStatusType.Crashed" --include=*.cs .` returns only the enum definition — nothing in the server ever assigns it. The status and its red dashboard styling exist; the code to set it doesn't.

There's also no config key anywhere for restarting a failed module (`grep -i restart` across `appsettings.json` and `ServerOptions.cs` returns nothing), which matches the report above that no auto-restart setting can be found.

## Why Coral in particular

As far as I can tell from the code, nothing here is Coral-specific — any module that exits unexpectedly is handled identically, because the handler is generic.

I have two deaths of this module logged, and they are **not** the same signature. Worth separating them.

**Spontaneous.** This is the one that matches the original report. From the server's point of view, nothing but the exit:

```
15:36:36:Response rec'd from Object Detection (Coral) command 'detect' (...3ebcde) ['No objects found']  took 67ms
15:36:36:Response rec'd from Object Detection (Coral) command 'detect' (...1b0673) ['No objects found']  took 64ms
15:37:24:Module ObjectDetectionCoral has shutdown
15:37:24:objectdetection_coral_adapter.py: has exited
```

Note that **both** of those last two lines come from the server, not the module. `has shutdown` is `ModuleExecutionComplete`, and `has exited` is synthesised in `SendOutputToLog` when the stdout stream closes empty:

```csharp
if (process.HasExited && string.IsNullOrEmpty(message))
    message = "has exited";
```

So the module emitted nothing at all. "Stopping without any logging" in the original report is literally accurate — but not because the module is being unhelpful. Windows' Application event log has the reason:

```
2026-09-04 15:37:22  [1000]  Application Error
  Faulting application name: python.exe, version: 3.9.6150.1013
  Faulting module name: libusb-1.0.dll, version: 1.0.26.11724
  Exception code: 0xc0000005
  Fault offset: 0x000000000000a241
  Faulting application path: C:\Program Files\CodeProject\AI\runtimes\bin\windows\python39\pyt...
```

`0xc0000005` is an access violation inside `libusb-1.0.dll` — the process was terminated by the OS on the spot. No Python exception, no glog fatal handler, no stdout flush. That is why there is nothing to find in the module's output: there was never an opportunity to write any.

This is a different failure from the unplug case below, even though both involve USB. Unplugging is *detected* by libedgetpu, which calls its own fatal handler, and that handler logs before aborting. A segfault inside libusb skips all of that.

Two things this rules out for my spontaneous failure, in case they're worth eliminating for others too:

- The device did not drop off the bus. `Microsoft-Windows-Kernel-PnP/Configuration` has no events in a four-hour window around the crash, `setupapi.dev.log` shows no device reconfiguration, and when I checked Device Manager ~18 hours later, before restarting anything, the Coral was still enumerated normally as `Coral USB Accelerator`.
- It wasn't USB selective suspend. The root hub the accelerator sits on reports `power-off allowed: False`.

I still don't know what triggered the access violation, and a libusb-level fault may well be out of scope for this repo. But it does sharpen the argument for the server-side change: **a process killed by an access violation cannot log, cannot clean up, and cannot restart itself.** Recovery has to come from the supervisor, and there isn't one.

Full WER signature, in case the offset means something to anyone: `libusb-1.0.dll` **1.0.26.11724**, exception `c0000005`, offset `0x000000000000a241`, event type `APPCRASH`. Note that `libusb-1.0.dll` and `edgetpu.dll` both load from `C:\WINDOWS\SYSTEM32\` — they come from Google's Edge TPU runtime installer, not from this repo. WER kept metadata only, no dump, so I can't resolve that offset to a function.

For contrast, the unplug crash below has a completely different signature: `ucrtbase.dll`, exception `c0000409` with data `0x7` (`FAST_FAIL_FATAL_APP_EXIT`) — ie a deliberate `abort()`, which is what glog's fatal handler does *after* it has written its message. That's the mechanical reason one crash logs and the other doesn't.

**Induced by unplugging the device.** Here the module does log, at fatal level:

```
09:36:17:Response rec'd from Object Detection (Coral) command 'detect' (...07af2a) ['No objects found']  took 67ms
09:37:24:objectdetection_coral_adapter.py: F driver/usb/usb_driver.cc:1148] HandleQueuedBulkIn transfer in failed. Unknown: USB transfer error 1 [LibUsbDataInCallback]
09:37:27:Module ObjectDetectionCoral has shutdown
09:37:27:objectdetection_coral_adapter.py: has exited
```

That's libedgetpu's own USB driver at glog fatal level (`F`), so the process aborts rather than raising anything Python could catch — which is why there's no traceback even here. A similar sequence has been reported on a different project ([[skrashevich/hassio-addons#17](https://github.com/skrashevich/hassio-addons/issues/17)](https://github.com/skrashevich/hassio-addons/issues/17)).

Both crashes involve USB, but they are not the same failure and they are not interchangeable evidence. One is a detected transfer error handled by libedgetpu; the other is a memory access violation inside libusb. What they have in common is the part this issue is about: the server treats both as a clean shutdown and recovers from neither.

### Reproduction

The induced crash above is repeatable on demand:

1. Start the Coral object detection module and confirm it's serving `detect` requests normally.
2. Physically unplug the Coral USB Accelerator.
3. The module logs the fatal libusb error and exits within a few seconds.
4. The server logs `** Module ObjectDetectionCoral has shutdown` at information level, sets the status to `Stopped`, and takes no further action.
5. Replug the device. Nothing restarts the module; it stays `Stopped` until restarted by hand.

**What this does and doesn't reproduce.** Steps 4 and 5 — the server treating a crash as a clean stop and never recovering — are reproduced exactly, and that's the part the patch addresses. It does **not** reproduce the silent death: unplugging produces a clear error, whereas the spontaneous failure produces none. So this is a reliable way to exercise the supervision gap, not a reproduction of whatever is killing the module unprompted.

Whether the USB link should drop at all is a separate hardware question. My accelerator has previously enumerated as `Unknown USB Device (Device Descriptor Request Failed)` and has sat in `DFU` mode after a reseat until the module was restarted, so it is demonstrably flaky. Making that link bulletproof is out of scope here. Surviving the drop isn't.

## Minor: what does the accelerator label in the dashboard mean?

Not part of the above, and I may be misreading it, but it confused me for a while so I'll mention it.

Immediately before the crash quoted above, the dashboard showed `CPU (TF-Lite)` for the module — yet the module was clearly holding an open USB connection to the TPU, since unplugging the device is what killed it. So the label didn't reflect what the module was actually using at that moment.

Is that field meant to report the active accelerator, or something else? If it is meant to report the active accelerator, it would be a useful thing to be able to trust — the `TPU detected` / `Using Edge TPU` lines only appear in the module's startup log, and their absence isn't something you'd notice unless you already suspected a problem.

## Proposed patch

Full diff in the collapsed block at the bottom. Four files, ~170 lines, no new dependencies, and no change to behaviour when a module is stopped or restarted deliberately.

1. **`ModuleExecutionComplete` distinguishes expected from unexpected exits.** An exit is treated as expected if the status at the time was `Stopping`, `Restarting` or `Stopped` — all three are set by `StopProcess` / `RestartProcess` before the process goes away. Anything else is treated as a crash.
2. **Unexpected exits set `ProcessStatusType.Crashed`** (the existing enum value) and log at error level, including the exit code and how long the module had been running.
3. **Unexpected exits schedule a restart** with exponential backoff and a cap on consecutive attempts, so a module that's genuinely broken at startup won't spin. The counter resets once a module has run for `ModuleStableRunMinutes`, so a module that dies once every few days always gets a fresh set of retries rather than slowly exhausting them.
4. **`ModuleRunner` passes the process service a `Func<string, ModuleConfig?>` resolver**, so a restart can find the module's config without `ModuleProcessServices` taking ownership of the module collection. This mirrors the existing `OnModuleStateChange` callback pattern.

All behind `ServerOptions.RestartCrashedModules`, defaulting to `true`, with the backoff, attempt cap and stability window configurable in `appsettings.json`.

### Caveats on the patch

- **It is not compiled or tested.** I don't have a .NET toolchain to hand. It needs a build and a soak test before anyone takes it seriously.
- The edit I'd most want reviewed is the expected-exit predicate. I've assumed those three statuses cover every path that stops a process deliberately, based on reading `StopProcess` and `RestartProcess` — if there's a path that kills a process without first setting one of them, that would be misread as a crash and restarted.
- The restart is fire-and-forget (`_ = RestartCrashedModuleAsync(moduleId)`) so it doesn't block the `Exited` handler. Exceptions inside it are caught and logged rather than swallowed silently, but a second pair of eyes on that would be sensible.
- With the default `MaxModuleRestartAttempts` of 5 and a 5s base, the backoff runs 5/10/20/40/80s, so the 120s ceiling only bites if you raise the attempt cap.

## Possibly related: server memory after a module dies

Flagging this separately because I have **not** diagnosed it and it may be nothing.

About 22 hours after the Coral module died, with the module still showing Stopped, `CodeProject.AI.Server` was holding ~2.6 GB working set.

Blue Iris kept sending `detect` requests throughout that window — its log shows over 1,100 cancelled alerts on a single camera and hundreds of `AI: not responding` entries, so the client was not backing off.

Reading `QueueServices`, the obvious leak isn't there: the channels are bounded (`MaxQueueLength`, 1024 in `appsettings.json`) and `_pendingResponses` is cleaned up in a `finally`. But requests are written to the channel with

```csharp
await queue.Writer.WriteAsync(request, theToken).ConfigureAwait(false);
```

and **as far as I can tell from reading the code**, nothing removes a request when the caller times out — only a dequeue does that. With no module consuming the queue, I'd expect up to 1024 requests, each carrying an image payload, to stay resident, with later writes failing as "queue is full" rather than displacing them.

That's a hypothesis about a mechanism, not a measured cause, and I don't think it accounts for 2.6 GB on its own — it may simply be .NET holding freed heap rather than returning it to the OS. I'm mentioning it only because "module dead" and "server memory high" co-occurred on my machine, and because draining a queue whose consumer has gone away seems reasonable regardless of whether it's the cause here.

---

<details>
<summary><b>Proposed patch (uncompiled — click to expand)</b></summary>

```diff
diff --git a/src/server/Config/ServerOptions.cs b/src/server/Config/ServerOptions.cs
index f195596..5be7db3 100644
--- a/src/server/Config/ServerOptions.cs
+++ b/src/server/Config/ServerOptions.cs
@@ -33,6 +33,37 @@ namespace CodeProject.AI.Server
         /// Gets or sets the environment variables, common to the CodeProject.AI Server ecosystem.
         /// </summary>
         public Dictionary<string, object>? EnvironmentVariables { get; set; }
+
+        /// <summary>
+        /// Gets or sets whether the server should automatically restart a module whose process
+        /// exits unexpectedly (ie a crash, rather than a shutdown or restart we asked for).
+        /// </summary>
+        public bool RestartCrashedModules { get; set; } = true;
+
+        /// <summary>
+        /// Gets or sets the maximum number of consecutive restart attempts made for a module that
+        /// keeps exiting unexpectedly. The count is reset once a module has run without incident
+        /// for ModuleStableRunMinutes.
+        /// </summary>
+        public int MaxModuleRestartAttempts { get; set; } = 5;
+
+        /// <summary>
+        /// Gets or sets the delay, in seconds, before the first restart attempt. Subsequent
+        /// attempts back off exponentially up to ModuleRestartMaxDelaySec.
+        /// </summary>
+        public int ModuleRestartBaseDelaySec { get; set; } = 5;
+
+        /// <summary>
+        /// Gets or sets the ceiling, in seconds, for the exponential restart backoff.
+        /// </summary>
+        public int ModuleRestartMaxDelaySec { get; set; } = 120;
+
+        /// <summary>
+        /// Gets or sets how long, in minutes, a module must run before it is considered stable. A
+        /// module that exits after running at least this long has its consecutive-crash counter
+        /// reset, so a module that dies once every few days is always given a fresh set of retries.
+        /// </summary>
+        public int ModuleStableRunMinutes { get; set; } = 5;
     }
 
     /// <summary>
diff --git a/src/server/Modules/ModuleProcessServices.cs b/src/server/Modules/ModuleProcessServices.cs
index 65230aa..ab970bb 100644
--- a/src/server/Modules/ModuleProcessServices.cs
+++ b/src/server/Modules/ModuleProcessServices.cs
@@ -28,6 +28,10 @@ namespace CodeProject.AI.Server.Modules
         private static readonly ConcurrentDictionary<string, ProcessStatus> _processStatuses  = new();
         private static readonly ConcurrentDictionary<string, Process>       _runningProcesses = new();
 
+        // Number of times, in a row, a module's process has exited unexpectedly. Reset once the
+        // module has run without incident for ServerOptions.ModuleStableRunMinutes.
+        private static readonly ConcurrentDictionary<string, int>           _consecutiveCrashes = new();
+
         private readonly VersionConfig                  _versionConfig;
         private readonly ServerOptions                  _serverOptions;
         private readonly QueueServices                  _queueServices;
@@ -73,6 +77,13 @@ namespace CodeProject.AI.Server.Modules
         /// </summary>
         public Func<ModuleConfig, Task>? OnModuleStateChange { get; set; } = null;
 
+        /// <summary>
+        /// A callback that resolves a module ID to its ModuleConfig. Set by ModuleRunner, which
+        /// owns the collection of installed modules. Needed so that a module which exits
+        /// unexpectedly can be restarted without this class having to own the module collection.
+        /// </summary>
+        public Func<string, ModuleConfig?>? GetModuleConfig { get; set; } = null;
+
         /// <summary>
         /// Gets the environment variables applied to all processes.
         /// </summary>
@@ -805,15 +816,118 @@ namespace CodeProject.AI.Server.Modules
                     return;
                 }
 
-                TryGetProcessStatus(moduleId, out ProcessStatus? processStatus);           
+                TryGetProcessStatus(moduleId, out ProcessStatus? processStatus);
+
+                // Capture the status *before* we overwrite it. If we (the server) asked this
+                // process to stop then it will already be Stopping or Restarting. Anything else
+                // means the process went away on its own: a crash, a native fault in a shared
+                // library, or the OS killing it. Those are the cases we want to recover from.
+                ProcessStatusType priorStatus = processStatus?.Status ?? ProcessStatusType.Unknown;
+                bool expectedExit = priorStatus == ProcessStatusType.Stopping   ||
+                                    priorStatus == ProcessStatusType.Restarting ||
+                                    priorStatus == ProcessStatusType.Stopped;
+
+                // How long did it run for? A module that ran for a good while before dying gets a
+                // fresh set of retries; one that dies immediately on start does not.
+                TimeSpan? uptime = processStatus?.Started is null
+                                 ? null
+                                 : DateTime.UtcNow - processStatus.Started.Value;
+
+                // ProcessStatusType.Crashed and its (red) dashboard styling already exist, but
+                // nothing has ever set it: until now every exit was reported as a clean stop.
                 if (processStatus is not null)
-                    processStatus.Status = ProcessStatusType.Stopped;
-
-                _logger.LogInformation($"** Module {moduleId} has shutdown");
+                    processStatus.Status = expectedExit ? ProcessStatusType.Stopped
+                                                        : ProcessStatusType.Crashed;
 
                 // Remove this from the list of running processes
                 if (_runningProcesses.TryGetValue(moduleId, out _))
                     _runningProcesses.TryRemove(moduleId, out _);
+
+                if (expectedExit)
+                {
+                    _logger.LogInformation($"** Module {moduleId} has shutdown");
+                    _consecutiveCrashes.TryRemove(moduleId, out _);
+                    return;
+                }
+
+                // An unexpected exit. Say so plainly, and include the exit code: previously this
+                // was logged identically to a clean shutdown, which made crashes very hard to spot.
+                int? exitCode = null;
+                try
+                {
+                    exitCode = process.ExitCode;
+                }
+                catch
+                {
+                    // ExitCode can throw if the process was never started successfully
+                }
+
+                _logger.LogError($"** Module {moduleId} exited unexpectedly" +
+                                 (exitCode is null ? string.Empty : $" (exit code {exitCode})") +
+                                 (uptime  is null ? string.Empty : $" after running for {uptime.Value.TotalMinutes:F1} mins"));
+
+                if (uptime is not null &&
+                    uptime.Value >= TimeSpan.FromMinutes(_serverOptions.ModuleStableRunMinutes))
+                {
+                    _consecutiveCrashes.TryRemove(moduleId, out _);
+                }
+
+                // Fire and forget: we must not block the Process.Exited event handler.
+                _ = RestartCrashedModuleAsync(moduleId);
+            }
+        }
+
+        /// <summary>
+        /// Restarts a module whose process exited unexpectedly, backing off exponentially and
+        /// giving up after a configurable number of consecutive failures.
+        /// </summary>
+        /// <param name="moduleId">The ID of the module to restart</param>
+        private async Task RestartCrashedModuleAsync(string moduleId)
+        {
+            if (!_serverOptions.RestartCrashedModules)
+                return;
+
+            ModuleConfig? module = GetModuleConfig?.Invoke(moduleId);
+            if (module is null || !module.Valid)
+            {
+                _logger.LogWarning($"Unable to resolve module {moduleId}, so it will not be restarted");
+                return;
+            }
+
+            int attempt = _consecutiveCrashes.AddOrUpdate(moduleId, 1, (_, count) => count + 1);
+            if (attempt > _serverOptions.MaxModuleRestartAttempts)
+            {
+                _logger.LogError($"Module {moduleId} has exited unexpectedly {attempt} times in a row. " +
+                                  "No further restarts will be attempted. Restart it from the " +
+                                  "dashboard once the underlying problem is resolved.");
+                return;
+            }
+
+            int delaySecs = Math.Min(_serverOptions.ModuleRestartBaseDelaySec * (1 << (attempt - 1)),
+                                     _serverOptions.ModuleRestartMaxDelaySec);
+
+            _logger.LogInformation($"Restarting module {moduleId} in {delaySecs}s " +
+                                   $"(attempt {attempt} of {_serverOptions.MaxModuleRestartAttempts})");
+
+            await Task.Delay(TimeSpan.FromSeconds(delaySecs)).ConfigureAwait(false);
+
+            // Guard against a race: the user may have started the module by hand while we waited.
+            if (_runningProcesses.ContainsKey(moduleId))
+            {
+                _logger.LogInformation($"Module {moduleId} is already running again. Restart abandoned.");
+                return;
+            }
+
+            try
+            {
+                if (await StartProcess(module, null).ConfigureAwait(false))
+                    _logger.LogInformation($"Module {moduleId} restarted successfully");
+                else
+                    _logger.LogError($"Failed to restart module {moduleId}");
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Error restarting module {moduleId}: {ex.Message}");
             }
         }
 
diff --git a/src/server/Modules/ModuleRunner.cs b/src/server/Modules/ModuleRunner.cs
index 5f66e61..b38729a 100644
--- a/src/server/Modules/ModuleRunner.cs
+++ b/src/server/Modules/ModuleRunner.cs
@@ -95,6 +95,12 @@ namespace CodeProject.AI.Server.Modules
             _meshMonitor      = meshMonitor;
             _logger           = logger;
 
+            // The process service needs to be able to look up a module by ID so that it can
+            // restart one whose process has exited unexpectedly. We own the module collection,
+            // so we hand it a resolver rather than duplicating that ownership.
+            ProcessService.GetModuleConfig = (moduleId) =>
+                _installedModules.TryGetValue(moduleId, out ModuleConfig? module) ? module : null;
+
             _logger.LogInformation($"** Server version:   {_versionConfig.VersionInfo!.Version}");
 #if DEBUG
             // Create a modules.json file each time we run
diff --git a/src/server/appsettings.json b/src/server/appsettings.json
index 8bd3450..865d4f5 100644
--- a/src/server/appsettings.json
+++ b/src/server/appsettings.json
@@ -58,6 +58,22 @@
     // Whether or not to disable IPv6 support.
     "DisableIPv6": false,
 
+    // Automatically restart a module whose process exits unexpectedly. This does not apply to
+    // modules that we stopped or restarted deliberately.
+    "RestartCrashedModules": true,
+
+    // How many times in a row we'll try to restart a module that keeps exiting unexpectedly
+    // before giving up. The counter resets once a module has run for ModuleStableRunMinutes.
+    "MaxModuleRestartAttempts": 5,
+
+    // Delay before the first restart attempt. Later attempts back off exponentially (5, 10, 20,
+    // 40, 80...) up to ModuleRestartMaxDelaySec.
+    "ModuleRestartBaseDelaySec": 5,
+    "ModuleRestartMaxDelaySec": 120,
+
+    // How long a module must run before we consider it healthy again and reset its crash count.
+    "ModuleStableRunMinutes": 5,
+
     // URLs and Endpoints on CodeProject.com
     // "ServerVersionCheckUrl": "https://www.codeproject.com/ai/version.aspx", // Check for latest version
     // "ServerDownloadUrl":     "https://www.codeproject.com/ai/latest.aspx",  // Link to download latest server
```

</details>

**Comment by melyux** [comment]
TL;DR. Tell the AI to cut it down next time

**Comment by ChangoMunk** [comment]
Summary for anyone skimming: ModuleRunner.ExecuteAsync starts modules once then sleeps on Task.Delay(Timeout.Infinite). ModuleExecutionComplete handles every exit identically — sets Stopped, logs at info level, no restart, exit code never read. ProcessStatusType.Crashed exists in the SDK and is styled red in dashboard.js but is never assigned anywhere in the server. Result: a module that dies stays dead and the log looks like a clean shutdown. Repro: unplug the Coral mid-run. Patch in the collapsed block above.
